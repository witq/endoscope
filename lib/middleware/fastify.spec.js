const fastifyPlugin = require("./fastify.js");

describe("fastifyMiddleware", () => {
  const fastifyInstance = {
    route: null
  };

  beforeEach(() => (fastifyInstance.route = jest.fn()));

  it("should register a plugin in fastify", () => {
    const plugin = fastifyPlugin({});
    const next = jest.fn();

    plugin(fastifyInstance, {}, next);

    expect(next).toBeCalled();
    expect(fastifyInstance.route.mock.calls).toMatchSnapshot();
    expect(fastifyInstance.route.mock.calls[0].handler).toBe(
      fastifyInstance.route.mock.calls[1].handler
    );
  });

  it("should respect the prefix option", () => {
    const plugin = fastifyPlugin({});

    plugin(
      fastifyInstance,
      {
        endoscope: {
          prefix: "/"
        }
      },
      () => {}
    );

    expect(fastifyInstance.route.mock.calls).toMatchSnapshot();
  });

  it("should respect the successCode option", () => {
    const endoscopeMock = {
      run: jest.fn(() => Promise.resolve([]))
    };
    const replyMock = {
      code: jest.fn()
    };
    const plugin = fastifyPlugin(endoscopeMock);

    plugin(
      fastifyInstance,
      {
        endoscope: {
          successCode: 200
        }
      },
      () => {}
    );

    const handler = fastifyInstance.route.mock.calls[0][0].handler;

    expect.assertions(1);

    return handler({ params: {} }, replyMock).then(() => {
      expect(replyMock.code).toBeCalledWith(200);
    });
  });

  it("should respect the errorCode option", () => {
    const endoscopeMock = {
      run: jest.fn(() => Promise.reject(""))
    };
    const replyMock = {
      code: jest.fn()
    };
    const plugin = fastifyPlugin(endoscopeMock);

    plugin(
      fastifyInstance,
      {
        endoscope: {
          errorCode: 666
        }
      },
      () => {}
    );

    const handler = fastifyInstance.route.mock.calls[0][0].handler;

    expect.assertions(1);

    return handler({ params: {} }, replyMock).then(() => {
      expect(replyMock.code).toBeCalledWith(666);
    });
  });

  it("should return succesful result", () => {
    const endoscopeMock = {
      run: jest.fn(() => Promise.resolve(["ok"]))
    };
    const replyMock = {
      code: jest.fn()
    };
    const plugin = fastifyPlugin(endoscopeMock);

    plugin(
      fastifyInstance,
      {},
      () => {}
    );

    const handler = fastifyInstance.route.mock.calls[0][0].handler;

    expect.assertions(1);

    return handler({ params: {} }, replyMock).then(result => {
      expect(result).toEqual(["ok"]);
    });
  });

  it("should return error result", () => {
    const endoscopeMock = {
      run: jest.fn(() => Promise.reject("err"))
    };
    const replyMock = {
      code: jest.fn()
    };
    const plugin = fastifyPlugin(endoscopeMock);

    plugin(
      fastifyInstance,
      {},
      () => {}
    );

    const handler = fastifyInstance.route.mock.calls[0][0].handler;

    expect.assertions(1);

    return handler({ params: {} }, replyMock).then(result => {
      expect(result).toEqual("err");
    });
  });
});
