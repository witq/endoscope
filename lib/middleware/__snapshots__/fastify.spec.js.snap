// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fastifyMiddleware should register a plugin in fastify 1`] = `
Array [
  Array [
    Object {
      "handler": [Function],
      "method": "GET",
      "url": "/healthz/:level",
    },
  ],
  Array [
    Object {
      "handler": [Function],
      "method": "GET",
      "url": "/healthz",
    },
  ],
]
`;

exports[`fastifyMiddleware should respect the prefix option 1`] = `
Array [
  Array [
    Object {
      "handler": [Function],
      "method": "GET",
      "url": "/:level",
    },
  ],
  Array [
    Object {
      "handler": [Function],
      "method": "GET",
      "url": "/",
    },
  ],
]
`;
